File: scripts/_mini_yaml.py
"""Minimal YAML loader for the project seeds.

This loader supports the subset of YAML used by the seed files so the
build scripts work in offline environments without PyYAML.
"""
from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Iterable, Tuple


@dataclass
class _Line:
    indent: int
    content: str


def _iter_lines(text: str) -> Iterable[_Line]:
    for raw in text.splitlines():
        stripped = raw.strip()
        if not stripped or stripped.startswith("#"):
            continue
        indent = len(raw) - len(raw.lstrip(" "))
        if stripped.startswith("#"):
            continue
        # remove trailing comments keeping inline braces
        hash_index = stripped.find(" #")
        if hash_index != -1:
            stripped = stripped[:hash_index].rstrip()
        yield _Line(indent=indent, content=stripped)


def _parse_value(token: str) -> Any:
    token = token.strip()
    if token == "":
        return None
    if token.lower() == "true":
        return True
    if token.lower() == "false":
        return False
    if token.lower() in {"null", "none"}:
        return None
    if token[0:1] in {"'", '"'} and token[-1:] == token[0]:
        return token[1:-1]
    # inline list
    if token.startswith("[") and token.endswith("]"):
        inner = token[1:-1].strip()
        if not inner:
            return []
        return [_parse_value(part) for part in _split_items(inner)]
    # inline map
    if token.startswith("{") and token.endswith("}"):
        inner = token[1:-1].strip()
        result = {}
        if not inner:
            return result
        for item in _split_items(inner):
            key, value = _split_pair(item)
            result[key] = _parse_value(value)
        return result
    # numeric handling
    try:
        if "." in token:
            return float(token)
        return int(token)
    except ValueError:
        return token


def _split_items(text: str) -> Iterable[str]:
    items = []
    current = []
    depth = 0
    quote: str | None = None
    escape = False
    for ch in text:
        if quote:
            current.append(ch)
            if escape:
                escape = False
                continue
            if ch == "\\":
                escape = True
                continue
            if ch == quote:
                quote = None
            continue
        if ch in {'"', "'"}:
            quote = ch
            current.append(ch)
            continue
        if ch in "[{":
            depth += 1
            current.append(ch)
            continue
        if ch in "]}":
            depth -= 1
            current.append(ch)
            continue
        if ch == "," and depth == 0:
            items.append("".join(current).strip())
            current = []
            continue
        current.append(ch)
    if current:
        items.append("".join(current).strip())
    return [item for item in items if item]


def _split_pair(item: str) -> Tuple[str, str]:
    if ":" not in item:
        raise ValueError(f"Invalid mapping token: {item}")
    key, value = item.split(":", 1)
    return key.strip().strip('"').strip("'"), value.strip()


def _parse_block(lines: Iterable[_Line], start: int, indent: int) -> Tuple[Any, int]:
    entries = list(lines)
    result: dict[str, Any] = {}
    i = start
    while i < len(entries):
        line = entries[i]
        if line.indent < indent:
            break
        if line.indent > indent:
            raise ValueError(f"Unexpected indentation at line: {line.content}")
        if line.content.startswith("- "):
            raise ValueError("Unexpected list item without key")
        key, value = _split_pair(line.content) if ":" in line.content else (line.content, "")
        key = key.strip()
        if value == "":
            # Determine if the block is a list or dict by peeking ahead
            sub_start = i + 1
            while sub_start < len(entries) and entries[sub_start].indent <= indent:
                if entries[sub_start].indent == indent:
                    break
                sub_start += 1
            j = i + 1
            # peek next meaningful line
            while j < len(entries) and entries[j].indent <= indent:
                j += 1
            next_line = entries[i + 1] if i + 1 < len(entries) else None
            if next_line and next_line.content.startswith("- ") and next_line.indent > indent:
                value_obj, new_index = _parse_list(entries, i + 1, indent + 2)
            else:
                value_obj, new_index = _parse_block(entries, i + 1, indent + 2)
            result[key] = value_obj
            i = new_index
        else:
            result[key] = _parse_value(value)
            i += 1
    return result, i


def _parse_list(entries: list[_Line], start: int, indent: int) -> Tuple[list[Any], int]:
    items: list[Any] = []
    i = start
    while i < len(entries):
        line = entries[i]
        if line.indent < indent:
            break
        if not line.content.startswith("- "):
            break
        token = line.content[2:].strip()
        if token == "":
            # nested block item
            value, new_index = _parse_block(entries, i + 1, indent + 2)
            items.append(value)
            i = new_index
            continue
        if token.startswith("{") and token.endswith("}"):
            value = _parse_value(token)
            items.append(value)
            i += 1
            continue
        if token.startswith("[") and token.endswith("]"):
            items.append(_parse_value(token))
            i += 1
            continue
        # simple scalar
        items.append(_parse_value(token))
        i += 1
    return items, i


def safe_load(stream: Any) -> Any:
    text = stream.read() if hasattr(stream, "read") else str(stream)
    lines = list(_iter_lines(text))
    if not lines:
        return {}
    result, index = _parse_block(lines, 0, lines[0].indent)
    if index < len(lines):
        # The first block may not cover everything if indentation resets.
        extra, _ = _parse_block(lines, index, lines[index].indent)
        result.update(extra)
    return result

File: scripts/_mini_yaml.ts
// RESUME_MARKER: CHK_1
/* eslint-disable @typescript-eslint/no-explicit-any */

type YAMLValue = any;

type Line = {
  indent: number;
  content: string;
};

function iterLines(text: string): Line[] {
  const result: Line[] = [];
  for (const raw of text.split(/\r?\n/)) {
    const stripped = raw.trim();
    if (!stripped || stripped.startsWith('#')) {
      continue;
    }
    const indent = raw.length - raw.replace(/^\s+/, '').length;
    let content = raw.slice(indent).trim();
    if (content.startsWith('#')) {
      continue;
    }
    const hashIndex = content.indexOf(' #');
    if (hashIndex !== -1) {
      content = content.slice(0, hashIndex).trimEnd();
    }
    if (!content) {
      continue;
    }
    result.push({ indent, content });
  }
  return result;
}

function parseValue(token: string): YAMLValue {
  const t = token.trim();
  if (t === '') {
    return null;
  }
  const lower = t.toLowerCase();
  if (lower === 'true') return true;
  if (lower === 'false') return false;
  if (lower === 'null' || lower === 'none') return null;
  if ((t.startsWith("'") && t.endsWith("'")) || (t.startsWith('"') && t.endsWith('"'))) {
    return t.slice(1, -1);
  }
  if (t.startsWith('[') && t.endsWith(']')) {
    const inner = t.slice(1, -1).trim();
    if (!inner) return [];
    return splitItems(inner).map(parseValue);
  }
  if (t.startsWith('{') && t.endsWith('}')) {
    const inner = t.slice(1, -1).trim();
    const obj: Record<string, YAMLValue> = {};
    if (!inner) return obj;
    for (const item of splitItems(inner)) {
      const [key, value] = splitPair(item);
      obj[key] = parseValue(value);
    }
    return obj;
  }
  if (/^[+-]?\d+$/.test(t)) {
    const intVal = Number.parseInt(t, 10);
    if (!Number.isNaN(intVal)) {
      return intVal;
    }
  }
  if (/^[+-]?(\d+\.\d*|\.\d+)([eE][+-]?\d+)?$/.test(t) || /^[+-]?\d+[eE][+-]?\d+$/.test(t)) {
    const floatVal = Number.parseFloat(t);
    if (!Number.isNaN(floatVal)) {
      return floatVal;
    }
  }
  return t;
}

function splitItems(text: string): string[] {
  const items: string[] = [];
  let current = '';
  let depth = 0;
  let quote: string | null = null;
  let escape = false;
  for (const ch of text) {
    if (quote) {
      current += ch;
      if (escape) {
        escape = false;
        continue;
      }
      if (ch === '\\') {
        escape = true;
        continue;
      }
      if (ch === quote) {
        quote = null;
      }
      continue;
    }
    if (ch === '"' || ch === "'") {
      quote = ch;
      current += ch;
      continue;
    }
    if (ch === '[' || ch === '{') {
      depth += 1;
      current += ch;
      continue;
    }
    if (ch === ']' || ch === '}') {
      depth -= 1;
      current += ch;
      continue;
    }
    if (ch === ',' && depth === 0) {
      if (current.trim()) {
        items.push(current.trim());
      }
      current = '';
      continue;
    }
    current += ch;
  }
  if (current.trim()) {
    items.push(current.trim());
  }
  return items;
}

function splitPair(item: string): [string, string] {
  const idx = item.indexOf(':');
  if (idx === -1) {
    throw new Error(`Invalid mapping token: ${item}`);
  }
  const key = item
    .slice(0, idx)
    .trim()
    .replace(/^['"]|['"]$/g, '');
  const value = item.slice(idx + 1).trim();
  return [key, value];
}

type ParseResult = [YAMLValue, number];

type LineList = Line[];

function parseBlock(entries: LineList, start: number, indent: number): ParseResult {
  const result: Record<string, YAMLValue> = {};
  let i = start;
  while (i < entries.length) {
    const line = entries[i];
    if (line.indent < indent) break;
    if (line.indent > indent) {
      throw new Error(`Unexpected indentation at line: ${line.content}`);
    }
    if (line.content.startsWith('- ')) {
      throw new Error('Unexpected list item without key');
    }
    const hasColon = line.content.includes(':');
    const [rawKey, rawValue] = hasColon ? splitPair(line.content) : [line.content.trim(), ''];
    if (rawValue === '') {
      const next = entries[i + 1];
      if (next && next.indent > indent && next.content.startsWith('- ')) {
        const [value, newIndex] = parseList(entries, i + 1, indent + 2);
        result[rawKey] = value;
        i = newIndex;
        continue;
      }
      const [value, newIndex] = parseBlock(entries, i + 1, indent + 2);
      result[rawKey] = value;
      i = newIndex;
    } else {
      result[rawKey] = parseValue(rawValue);
      i += 1;
    }
  }
  return [result, i];
}

function parseList(entries: LineList, start: number, indent: number): ParseResult {
  const items: YAMLValue[] = [];
  let i = start;
  while (i < entries.length) {
    const line = entries[i];
    if (line.indent < indent) {
      break;
    }
    if (!line.content.startsWith('- ')) {
      break;
    }
    const token = line.content.slice(2).trim();
    if (token === '') {
      const [value, newIndex] = parseBlock(entries, i + 1, indent + 2);
      items.push(value);
      i = newIndex;
      continue;
    }
    if ((token.startsWith('{') && token.endsWith('}')) || (token.startsWith('[') && token.endsWith(']'))) {
      items.push(parseValue(token));
      i += 1;
      continue;
    }
    items.push(parseValue(token));
    i += 1;
  }
  return [items, i];
}

export function safeLoad(input: string | { read(): string }): YAMLValue {
  const text = typeof input === 'string' ? input : input.read();
  const lines = iterLines(text);
  if (lines.length === 0) {
    return {};
  }
  const [result, index] = parseBlock(lines, 0, lines[0]?.indent ?? 0);
  if (index < lines.length) {
    const nextIndent = lines[index]?.indent ?? 0;
    const [extra] = parseBlock(lines, index, nextIndent);
    if (typeof result === 'object' && result !== null && !Array.isArray(result)) {
      Object.assign(result, extra as Record<string, YAMLValue>);
    }
  }
  return result;
}

export default { safeLoad };
